Subject: [PATCH 1/1] scsi: scsi scan use after free fix                                                                                                              
From: brking@us.ibm.com                                                                                                                                              
To: olh@suse.de                                                                                                                                                      
Cc: benh@kernel.crashing.org, paulus@samba.org, brking@us.ibm.com                                                                                                    


Current scsi scanning code appears to have a use after free
bug is a LLDD's slave_alloc fails. Remove the redundant
scsi_free_queue.

...
>> Current scsi scanning code appears to have a use after free                                                                                                       
>> bug is a LLDD's slave_alloc fails. Remove the redundant                                                                                                           
>> scsi_free_queue.                                                                                                                                                  
>                                                                                                                                                                    
> Tested on a p710, works there.                                                                                                                                     
>                                                                                                                                                                    

Unfortunately, this exposes a memory leak in the cfq scheduler.
If you repeatedly:

echo "- - -" > /sys/class/scsi_host/host0/scan

you will soon run out of memory with this patch. Works fine with
the as scheduler.

Jens - do you have a bug report or more information regarding the
use after free problems introduced by my attempt to fix the dangling
reference in cfq, which was reverted?

http://marc.theaimsgroup.com/?l=git-commits-head&m=112614285132712&w=2
...

Signed-off-by: Brian King <brking@us.ibm.com>
Signed-off-by: Olaf Hering <olh@suse.de>

 drivers/scsi/scsi_scan.c |    1 -
 1 files changed, 1 deletion(-)

Index: linux-2.6.15-rc5-olh/drivers/scsi/scsi_scan.c
===================================================================
--- linux-2.6.15-rc5-olh.orig/drivers/scsi/scsi_scan.c
+++ linux-2.6.15-rc5-olh/drivers/scsi/scsi_scan.c
@@ -279,7 +279,6 @@ static struct scsi_device *scsi_alloc_sd
 
 out_device_destroy:
 	transport_destroy_device(&sdev->sdev_gendev);
-	scsi_free_queue(sdev->request_queue);
 	put_device(&sdev->sdev_gendev);
 out:
 	if (display_failure_msg)
